samza-log4j2/src/main/java/org/apache/samza/logging/log4j2/StreamAppender.java

```diff
@@ -152,18 +144,19 @@ public String getStreamName() {
   }
 
   /**
-   * Getter for the Config parameter.
+   * This should only be called after verifying that the {@link LoggingContextHolder} has the config.
```

Comment:
How do we ensure/enforce that?

Reply:
there is actually no tangible way to enforce it. Even prior to this PR, there were only checks to see if the AM's config is available via "JobModelManager.currentJobModelManager() != null" before using the config or calling this getConfig() and no real enforcement. But at that time, there was no explicit comment/doc stating this requirement/caution. now the same thing happens but via "this.loggingContextHolder.getConfig() != null".

the thing about StreamAppender is that because its a log appender but depends on non-logging components' flow and availability (for config..) its flow becomes inter-twined with those components  and hence hard to follow. For example, config get fetched after some component is up and config is asked for within setupSystem which sits as a no-op until config is avail. in other words, the internal methods within this class are dependent and called by other methods BASED on external components flow. This is what lends to this confusing pattern.

Reply:
This usage pattern constraint actually already exists in the current implementation. In the AM, it is already not guaranteed that the config is available. I was just adding some documentation to try to help clarify that now.
Unfortunately, to improve this, we would need to change the semantics of some other parts of this class. I actually did make some other changes in the initial iteration, but @lakshmi-manasa-g made a good point about maintaining compatibility within the scope of this PR. There is more discussion of this at https://github.com/apache/samza/pull/1532#discussion_r714217467.
Ideally, there would have been one fetch to the config and then that would have been passed around as a local argument instead of through an instance variable. That would have allowed better enforcement of the availability of the config. For now, this continues to need to be manually checked for ordering of method calls.

Reply:
Thanks for the details.